### PR TITLE
Backport: Add permissions check for secrets

### PR DESCRIPTION
**Pull request by rmweir**
Backport: https://github.com/rancher/rancher/pull/21052

### CODE REVIEW

File: pkg/api/customization/secret/validator.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+		// minimum info needed to use CanDo
+		secretState := map[string]interface{}{
+			"name":        data["name"],
+			"id":          id,
+			"namespaceId": data["namespaceId"],
+		}
+
+		// update is used here to avoid the application of general user permissions for secrets
+		if err := request.AccessControl.CanDo(v1.SecretGroupVersionKind.Group, v1.SecretResource.Name, "update", request, secretState, schema); err != nil {
+			return httperror.NewAPIError(httperror.PermissionDenied, fmt.Sprintf("unable to validate: user does not have permission"))
```
Comment (cjellick) [inline pkg/api/customization/secret/validator.go]: change "unable to validate: user does not have permission" to simply "unauthorized"

Comment (cjellick) [reply pkg/api/customization/secret/validator.go]: "unable to validate" just isnt a useful error to display

File: pkg/api/customization/secret/validator.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		}
+	} else if request.Method == http.MethodPut {
+		var secretState map[string]interface{}
+
+		if err := access.ByID(request, request.Version, request.Type, request.ID, &secretState); err != nil {
+			return httperror.NewAPIError(httperror.NotFound, fmt.Sprintf("unable to validate: %v", err))
```
Comment (cjellick) [inline pkg/api/customization/secret/validator.go]: you dont know this is a notFound error here.
Technically, the right thing to do would be to check to see if the error is a not found error and if it is, return NotFound, else return a 500 error with just the raw error message.

File: pkg/api/customization/secret/validator.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		}
+		// this is unused but will be necessary if readonly users are ever given permission to view secrets
+		if err := request.AccessControl.CanDo(v1.SecretGroupVersionKind.Group, v1.SecretResource.Name, "update", request, secretState, schema); err != nil {
+			return httperror.NewAPIError(httperror.PermissionDenied, "unable to validate: user does not have permission")
```
Comment (cjellick) [inline pkg/api/customization/secret/validator.go]: just say unauthorized

File: tests/integration/suite/test_rbac.py [hunk 4]
```diff
@@ -475,19 +476,147 @@ def test_member_can_perform_app_action(admin_mc, admin_pc, remove_resource,
...
+    prtb = admin_mc.client.create_project_role_template_binding(
+        name="prtb-" + random_str(),
+        userId=user_mc.user.id,
+        projectId=project.id,
+        roleTemplateId="read-only")
+    remove_resource(prtb)
+
+    wait_until_available(user_client, project)
+
+    proj_client = user_project_client(user_mc, project)
```
Comment (dramich) [inline tests/integration/suite/test_rbac.py]: Nit but I think this should be renamed to be clear it's the user client

File: pkg/api/customization/secret/validator.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+		// update is used here to avoid the application of general user permissions for secrets
+		if err := request.AccessControl.CanDo(v1.SecretGroupVersionKind.Group, v1.SecretResource.Name, "update", request, secretState, schema); err != nil {
+			return httperror.NewAPIError(httperror.PermissionDenied, "unauthorized")
+		}
+	} else if request.Method == http.MethodPut {
+		var secretState map[string]interface{}
+
+		if err := access.ByID(request, request.Version, request.Type, request.ID, &secretState); err != nil {
+			if httperror.IsNotFound(err) || isUnauthorized(err) {
```
Comment (dramich) [inline pkg/api/customization/secret/validator.go]: why do we 404 on unauthorized?

**Comment by dramich** [comment]
lgtm

File: pkg/api/customization/secret/validator.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	v1 "github.com/rancher/types/apis/core/v1"
+)
+
+func Validator(request *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
+	if request.Method == http.MethodPost {
+		id := ""
+
+		// extracting project name from data
+		if projectData, ok := data["projectId"].(string); ok {
+			if projectParts := strings.Split(projectData, ":"); len(projectParts) > 1 {
```
Comment (alena1108) [inline pkg/api/customization/secret/validator.go]: Should we return an error if by some reason len(projectParts) <=1?

**Comment by alena1108** [comment]
lgtm

**Pull request merged**